tasks/mc/data.py

```diff
@@ -435,7 +435,7 @@ class AllMCData(WrapperTask):
 
     def requires(self):
         for geography in [x.replace(' ', '_') for x in GEOGRAPHIES[self.country]]:
-            yield MCData(geography=geography)
+            yield MCData(geography=geography, country=self.country)
```

Comment:
I'd change this to return the array instead of yields:
```
return [MCData(geography=x.replace(' ', '_'), country=self.country) for x in GEOGRAPHIES[self.country]]`
```

This avoids the fun thing with Luigi and yield. Example with two yields:

1. Luigi tries to execute this task
2. requires() yields T1
3. Luigi checks that T1 is not complete and executes it
4. Luigi tries to execute this task
5. requires() yields T1
6. Luigi checks that T1 is complete
7. requires() yields T2
8. Luigi checks that T2 is not complete and executes it
9. Luigi tries to execute this task
10. requires() yields T1
11. Luigi checks that T1 is complete
12. requires() yields T2
13. Luigi checks that T2 is complete
14. This task is run

Also, it inhibits parallelism (since it won't get T2 until T1 is finished. If you return an array of tasks, this goes much better:
1. Luigi tries to execute this task
2. requires() returns [T1, T2]
3. Luigi checks that T1 is not complete and executes it
4. Luigi checks that T2 is not complete and executes it (in parallel with 3 if it has multiple workers)
5. Luigi tries to execute this task
6. requires() returns [T1, T2]
7. Luigi checks that T1 is complete
8. Luigi checks that T2 is complete
9. This task is run

Less calls to complete(), less calls to requires() and potential parallelism. Do not use `yield` unless absolutely necessary (i.e: which requirements you have depend on the result of some other requirement).

https://luigi.readthedocs.io/en/stable/tasks.html#dynamic-dependencies

> This mechanism is an alternative to Task.requires in case you are not able to build up the full dependency graph before running the task. It does come with some constraints: the Task.run method will resume from scratch each time a new task is yielded.

This mentions `Task.run`, but same applies to `yields` inside `requires` (which is not even documented I think).

Reply:
Same at AllMCCountries, right? Nice! Great explanation, I didn't know that and it'll be very helpful if parallelism is addressed.

Reply:
Yup, same problem there.